src/sync/app.cpp

```diff
@@ -0,0 +1,153 @@
+
+namespace realm {
+namespace app {
+
+std::shared_ptr<App> App::app(const std::string app_id,
+                              const realm::util::Optional<App::Config> config)
+{
+    return std::make_shared<App>(app_id, config);
+}
+
+
+
+static std::unique_ptr<error::AppError> handle_error(const Response response) {
+    if (response.body.empty()) {
+        return {};
+    }
+
+    if (response.headers.find("Content-Type") != response.headers.end()) {
+        if (response.headers.at("Content-Type") == "application/json") {
+            auto body = nlohmann::json::parse(response.body);
+            return std::make_unique<error::AppError>(app::error::ServiceError(body["errorCode"].get<std::string>(),
+                                  body["error"].get<std::string>()));
+        }
+    }
+
+    return {};
+}
+
+inline bool response_code_is_fatal(int status) {
+    // FIXME: we need more robust handling here
+    return status >= 300;
+}
+
+void App::login_with_credentials(const std::shared_ptr<AppCredentials> credentials,
+                                 int timeout_ms,
+                                 std::function<void(std::shared_ptr<SyncUser>, std::unique_ptr<error::AppError>)> completion_block) {
+    // construct the route
+    std::string route = util::format("%1/providers/%2/login", m_auth_route, provider_type_from_enum(credentials->m_provider));
+
+    auto handler = [&](const Response response) {
+        // if there is a already an error code, pass the error upstream
+        if (response_code_is_fatal(response.status_code)) { // FIXME: handle
+            return completion_block(nullptr, handle_error(response));
+        }
+
+        auto j = nlohmann::json::parse(response.body);
+
+        realm::SyncUserIdentifier identifier {
+            j["user_id"].get<std::string>(),
+            m_auth_route
+        };
+
+        std::shared_ptr<realm::SyncUser> sync_user;
+        try {
+            sync_user = realm::SyncManager::shared().get_user(identifier,
+                                                              j["refresh_token"].get<std::string>(),
+                                                              j["access_token"].get<std::string>());
+        } catch (const error::AppError& err) {
+            return completion_block(nullptr, std::make_unique<error::AppError>(err));
+        }
+
+        std::string profile_route = util::format("%1/auth/profile", m_base_route);
+        std::string bearer = util::format("Bearer %1", sync_user->access_token());
+
+        GenericNetworkTransport::get()->send_request_to_server({
+            Method::get,
+            profile_route,
+            timeout_ms,
+            {
+                { "Content-Type", "application/json;charset=utf-8" },
+                { "Accept", "application/json" },
+                { "Authorization", bearer}
+            },
+            std::string()
+        }, [&](const Response profile_response) {
```

Comment:
One use case in Java is having an API like:

```kotlin
val task: RealmAsyncTask = RealmApp.loginAsync("user", "password", Callback() {
  override fun onSuccess(user: RealmAppUser) { /* User logged in */ }
  override fun onError(error: ObjectServerError) { /* An error happend */ }
});

// Manually cancel. Callbacks should not trigger
task.cancel()
```

This looks kinda doable since the network request is being triggered in Java where we can cancel it. I guess we can wrap that exception with a Java-specific error code that will be ignored when parsed back to the completion block,.